src/librustc_typeck/check/method/confirm.rs

```diff
@@ -334,8 +340,18 @@ impl<'a, 'tcx> ConfirmContext<'a, 'tcx> {
         )
     }
 
-    fn unify_receivers(&mut self, self_ty: Ty<'tcx>, method_self_ty: Ty<'tcx>) {
-        match self.at(&self.misc(self.span), self.param_env).sup(method_self_ty, self_ty) {
+    fn unify_receivers(
+        &mut self,
+        self_ty: Ty<'tcx>,
+        method_self_ty: Ty<'tcx>,
+        pick: &probe::Pick<'tcx>,
+    ) {
+        debug!(
+            "unify_receivers: self_ty={:?} method_self_ty={:?} span={:?} pick={:?}",
+            self_ty, method_self_ty, self.span, pick
+        );
+        let cause = self.cause(self.span, ObligationCauseCode::UnifyReceiver(Rc::new(pick.item)));
```

Comment:
If you store `self.param_env` and `self_ty` in `UnifyReceiver`, then the diagnostic code can use `Instance::resolve` to determine the exact `impl` item being used.

Reply:
I tried a couple of things, without fruition, but I don't see how I can use `Instance::resolve` for this.

Reply:
@estebank: With the `DefId` of the associated item (a method), you can use `Instance::resolve` to determine the method actually being invoked. The substs should be available in the `confirm` method in this file.

Reply:
I tried that but the substs from `confirm` but it didn't work. I'll try something else soon.

Reply:
You might need to add the self type as the first item in the substa.

Reply:
You were right, I'd forgotten about the `self` subst, but now I'm encountering
```
error: internal compiler error:
src/librustc_trait_selection/traits/codegen/mod.rs:62:17:
Encountered error `Unimplemented` selecting `Binder(<&dyn Bar as Foo>)` during codegen
```

This makes me think that we won't be able to rely on that for this.